packages/llmRunner/src/llmRunner/LLMStep.ts

```diff
@@ -0,0 +1,235 @@
+  constructor(
+    public readonly name: string,
+    public readonly shape: Shape,
+    public readonly execute: (context: ExecuteContext<Shape>) => Promise<void>
+  ) {}
+
+  instantiate(
+    workflow: Workflow,
+    inputs: Inputs<Shape>
+  ): LLMStepInstance<Shape> {
+    return new LLMStepInstance(this, workflow, inputs);
+  }
+}
+
+export class LLMStepInstance<const Shape extends StepShape = StepShape> {
+  public durationMs?: number;
+  private logger: Logger;
+  private conversationMessages: Message[] = [];
+  public llmMetricsList: LlmMetrics[] = [];
+  private startTime: number;
+  private state: "PENDING" | "DONE" | "FAILED" = "PENDING";
+  private inputs: Inputs<Shape>;
+  private outputs: Partial<Outputs<Shape>> = {};
+
+  constructor(
+    public readonly template: LLMStepTemplate<Shape>,
+    public readonly workflow: Workflow,
+    inputs: Inputs<Shape>
+  ) {
+    this.startTime = Date.now();
+    this.logger = new Logger();
+    this.inputs = inputs;
+  }
+
+  getInput<K extends keyof Inputs<Shape>>(key: K): Inputs<Shape>[K] {
+    return this.inputs[key];
+  }
+
+  getLogs(): TimestampedLogEntry[] {
+    return this.logger.logs;
+  }
+
+  getConversationMessages(): Message[] {
+    return this.conversationMessages;
+  }
+
+  async run() {
+    if (this.state !== "PENDING") {
+      return;
+    }
+
+    const limits = this.workflow.checkResourceLimits();
+    if (limits) {
+      this.criticalError(limits);
+      return;
+    }
+
+    const executeContext: ExecuteContext<Shape> = {
+      getInput: (key) => this.getInput(key),
+      setOutput: (key, value) => this.setOutput(key, value),
+      log: (log) => this.log(log),
+      queryLLM: (promptPair) => this.queryLLM(promptPair),
+      workflow: this.workflow,
+    };
+
+    try {
+      await this.template.execute(executeContext);
+    } catch (error) {
+      this.criticalError(
+        error instanceof Error ? error.message : String(error)
+      );
+      return;
+    }
+    this.complete();
+  }
+
+  getState() {
+    return this.state;
+  }
+
+  getAllOutputs() {
+    return this.outputs;
+  }
+
+  getInputs() {
+    return this.inputs;
+  }
+
+  // private methods
+
+  private setOutput<K extends keyof Outputs<Shape>>(
+    key: K,
+    value: Outputs<Shape>[K]
+  ): void {
+    // TODO - check if output is already set?
+    this.outputs[key] = value;
+  }
+
+  private log(log: LogEntry): void {
+    this.logger.log(log);
+  }
+
+  private criticalError(error: string) {
+    this.log({ type: "error", message: error });
+    this.updateDuration();
+    this.state = "FAILED";
+  }
+
+  private updateDuration() {
+    this.durationMs = Date.now() - this.startTime;
+  }
+
+  private complete() {
+    this.updateDuration();
+    this.state = "DONE";
+  }
+
+  private addConversationMessage(message: Message): void {
+    this.conversationMessages.push(message);
+  }
+
+  private async queryLLM(promptPair: PromptPair): Promise<string | null> {
```

Comment:
Eventually we'll need other kinds of steps (e.g. for RAG), and then this method won't be appropriate on the step class.

But for now this seems fine; this class stores `conversationMessages` and `llmMetricsList`, so it's LLM-specific.